File: leetcode_75/Graphs_DFS/python/547.py
from typing import List


class Solution:
    def dfs(self, node, graph, visited):
        visited[node] = True
        for v in graph[node]:
            if not visited[v]:
                self.dfs(v, graph, visited)

    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        graph = {i: set() for i in range(1, n + 1)}
        for i in range(1, n + 1):
            for j in range(1, len(isConnected[i - 1]) + 1):
                if isConnected[i - 1][j - 1] and i != j:
                    graph[i].add(j)
                    graph[j].add(i)

        visited = [False] * (n + 1)
        count = 0
        for i in range(1, n + 1):
            if not visited[i]:
                self.dfs(i, graph, visited)
                count += 1

        return count

File: leetcode_75/Graphs_DFS/cpp/547.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution
{
public:
    int findCircleNum(vector<vector<int>> &isConnected)
    {
        int n = isConnected.size();
        unordered_map<int, vector<int>> graph;
        for (int i = 1; i < n + 1; ++i)
        {
            for (int j = 1; j < isConnected[i].size() + 1; ++j)
            {
                if (isConnected[i - 1][j - 1] && i != j)
                {
                    graph[i].push_back(j);
                    graph[j].push_back(i);
                }
            }
        }

        vector<bool> visited(n + 1, false);
        int count = 0;
        for (int i = 1; i < n + 1; ++i)
        {
            if (!visited[i])
            {
                this->dfs(i, graph, visited);
                count++;
            }
        }

        return count;
    }

private:
    void dfs(int node, unordered_map<int, vector<int>> &graph, vector<bool> &visited)
    {
        visited[node] = true;

        for (auto v : graph[node])
        {
            if (!visited[v])
            {
                this->dfs(v, graph, visited);
            }
        }
    }
};